partitioner: avoid inserting duplicates into heap

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/145081

This looks like it was a source of quadratic compile times in the torchtitan CP graphs. There's some code in the partitioner that iteratively adds users of a node to a heap, and pops the earliest user. If you have long parallel chains of fusible ops that all eventually feed into some shared ops, then this can result in:
(1) a node getting added to the heap many times
(2) each time we pop that node, we add (duplicates of) each of that node users to the heap
(3) repeat with each user

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145082



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh committed:
partitioner: avoid inserting duplicate node in heap

[ghstack-poisoned]

bdhirsh commented on benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher_partitioner2.py:
```diff
@@ -0,0 +1,58 @@
+    ]
+
+    for benchmark in all:
+        benchmark.enable_compile_time_instruction_count().collect_all().append_results(
```
I tested locally and this PR cuts the instruction-count of this microbenchmark from 61B -> 4B instructions (it grows a lot larger if you increase `len(tmps)` from 16 to 32)

bdhirsh commented on benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher_partitioner2.py:
```diff
@@ -0,0 +1,58 @@
+        f(self.x)
+
+
+def main():
```
hey @laithsakka - is there anything else I need to do to ensure this new compile time benchmark runs in CI? (do I need to update one of the expected-instruction-count files locally?

laithsakka replied:
replied privately but posting here for others 
(1) when you land this diff the benchmark will run, but it wont fail at regression .
(2) you have two benchmarks with the same name you probably want to rename it. 
(3) to enable failing at regressions you need to add a line to /data/users/lsakka/fbsource/fbcode/caffe2/benchmarks/dynamo/pr_time_benchmarks/expected_results.csv you can get the results from the logs [https://github.com/pytorch/pytorch/actions/runs/12832763265/job/35794624737](https://l.workplace.com/l.php?u=https%3A%2F%2Fgithub.com%2Fpytorch%2Fpytorch%2Factions%2Fruns%2F12832763265%2Fjob%2F35794624737&h=AT0YF52FqBqhPca4o3ysOT2Cau9R1jxM83wY68zV_jWiU2RwUSM3qL5-f3HNkYSlykqJdoVsDlBXlIPGbEdRh95Z77j8TBHsPoq7ix1Dovbp7ARaUDT_vNJf9jnJa2PczqIHG12tADuv7XxLPzhgwh1caVV8qf2gV0o-tg)

(4) usually i do 3 in different step, first land the diff that enable running the benchmark, then monitor it on https://fburl.com/unidash/vblyya4c to make sure its stable and not noisy for day or two then update the expected results  file above .

bdhirsh committed:
Update on "partitioner: avoid inserting duplicates into heap"

Fixes https://github.com/pytorch/pytorch/issues/145081

This looks like it was a source of quadratic compile times in the torchtitan CP graphs. There's some code in the partitioner that iteratively adds users of a node to a heap, and pops the earliest user. If you have long parallel chains of fusible ops that all eventually feed into some shared ops, then this can result in:
(1) a node getting added to the heap many times
(2) each time we pop that node, we add (duplicates of) each of that node users to the heap
(3) repeat with each user




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "partitioner: avoid inserting duplicates into heap"

Fixes https://github.com/pytorch/pytorch/issues/145081

This looks like it was a source of quadratic compile times in the torchtitan CP graphs. There's some code in the partitioner that iteratively adds users of a node to a heap, and pops the earliest user. If you have long parallel chains of fusible ops that all eventually feed into some shared ops, then this can result in:
(1) a node getting added to the heap many times
(2) each time we pop that node, we add (duplicates of) each of that node users to the heap
(3) repeat with each user




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "partitioner: avoid inserting duplicates into heap"

Fixes https://github.com/pytorch/pytorch/issues/145081

This looks like it was a source of quadratic compile times in the torchtitan CP graphs. There's some code in the partitioner that iteratively adds users of a node to a heap, and pops the earliest user. If you have long parallel chains of fusible ops that all eventually feed into some shared ops, then this can result in:
(1) a node getting added to the heap many times
(2) each time we pop that node, we add (duplicates of) each of that node users to the heap
(3) repeat with each user




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented on benchmarks/dynamo/pr_time_benchmarks/expected_results.csv:
```diff
@@ -18,11 +18,11 @@ add_loop_inductor_gpu,compile_time_instruction_count,27360000000,0.015
 
 
-basic_modules_ListOfLinears_inductor,compile_time_instruction_count,21310000000,0.015
+basic_modules_ListOfLinears_inductor,compile_time_instruction_count,19610000000,0.015
```
@laithsakka I updated the csv since [this job](https://github.com/pytorch/pytorch/actions/runs/12998305969/job/36259356615) reported a 7.95% win on this test.

It also reported a few regressions below, although they didn't actually fail the CI because they are within the tolerance. Can I assume that those differences could be from previous PRs, which moved the instruction counts a bit but did not require refreshing the expected results?

laithsakka replied:
if they are with in telerance you are ok

bdhirsh committed:
Update on "partitioner: avoid inserting duplicates into heap"

Fixes https://github.com/pytorch/pytorch/issues/145081

This looks like it was a source of quadratic compile times in the torchtitan CP graphs. There's some code in the partitioner that iteratively adds users of a node to a heap, and pops the earliest user. If you have long parallel chains of fusible ops that all eventually feed into some shared ops, then this can result in:
(1) a node getting added to the heap many times
(2) each time we pop that node, we add (duplicates of) each of that node users to the heap
(3) repeat with each user




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
